Source/ACE.Database/CachingWorldDatabase.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using ACE.Entity;
+using System.Collections.Concurrent;
+
+namespace ACE.Database
+{
+    public class CachingWorldDatabase : IWorldDatabase
+    {
+        /// <summary>
+        /// wrapper to the the world database that actually does the lifting
+        /// </summary>
+        private IWorldDatabase _wrappedDatabase;
+
+        /// <summary>
+        /// so, caches are fun.  we have to be especially careful that we never hand out the same instance
+        /// twice.  in order to do that, we always clone objects coming out of the cache and return the
+        /// clone instead of the actual cached object.  Concurrent dictionary could have been used, but
+        /// I went with a standard dictionary and a double locking system because the 99% use case is a
+        /// cache hit, and Currents don't do that as fast.
+        /// </summary>
+        private volatile Dictionary<uint, AceObject> _weenieCache = new Dictionary<uint, AceObject>();
+
+        private object _weenieCacheMutex = new object();
+
+        public CachingWorldDatabase(IWorldDatabase wrappedDatabase)
+        {
+            _wrappedDatabase = wrappedDatabase;
+        }
+
+        public AceObject GetAceObjectByWeenie(uint weenieClassId)
+        {
+            if (!_weenieCache.ContainsKey(weenieClassId))
```

Comment:
This is most likely unsafe.  Unless ContainsKey is guaranteed atomic w.r.t. _weenieCache (if Dictionaries are hash maps under the hood its highly unlikely), then this will eventually crash the code.  Double-check locking relies on a strictly atomic and ordered pre-check (volatility guarantees ordering).